### PR TITLE
feat: [CDS-47438]: TAS App setup steps

### PR DESCRIPTION
**Pull request by PratyushGar**
### Summary

- TAS file coverage will be added as part of this [ticket](https://harness.atlassian.net/browse/CDS-48123)

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TASSetupSource.tsx [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      <FormInput.RadioGroup
+        name="spec.tasInstanceCountType"
+        label={<Text>{getString('instanceFieldOptions.instanceText')}</Text>}
+        items={[
+          {
+            label: 'Read from Manifest',
+            value: InstancesType.FromManifest
+          },
+          {
+            label: 'Match Running Instances',
```
Comment (rameshx) [inline src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TASSetupSource.tsx:100]: add labels to strings

File: src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TasBasicAppSetupInputSet.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            template={template}
+            fieldPath={'spec.existingVersionToKeep'}
+          />
+        </div>
+      )}
+      {getMultiTypeFromValue((template?.spec as any)?.additionalRoutes) === MultiTypeInputType.RUNTIME && (
+        <div className={cx(stepCss.formGroup, stepCss.md)}>
+          <FormMultiTypeKVTagInput
+            name={`${path}.spec.additionalRoutes`}
+            tagsProps={{ placeholder: 'Type and press enter to add a route' }}
```
Comment (rameshx) [inline src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TasBasicAppSetupInputSet.tsx:77]: add to strings

File: src/modules/75-cd/components/PipelineSteps/TasBGAppSetup/TasBGAppSetupWidget.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+          }
+          return commonValidation.call(this, value, getString('version'))
+        }
+      }),
+      tempRoutes: Yup.lazy((value): Yup.Schema<unknown> => {
+        /* istanbul ignore else */
+        if (getMultiTypeFromValue(value as any) === MultiTypeInputType.FIXED) {
+          return Yup.array().min(1, 'Temp route is mandatory')
+        }
+        return Yup.string().required('Temp route is mandatory')
```
Comment (rameshx) [inline src/modules/75-cd/components/PipelineSteps/TasBGAppSetup/TasBGAppSetupWidget.tsx:81]: add to strings

File: src/modules/75-cd/components/PipelineSteps/TasCanaryAppSetup/TasCanaryAppSetupWidget.tsx [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            <div className={cx(stepCss.formGroup, stepCss.lg)}>
+              <FormInput.Select
+                items={[
+                  {
+                    label: 'Add new instances first, then downsize old instances',
+                    value: ResizeStrategyType.UpScaleNewFirst
+                  },
+                  {
+                    label: 'Downsize old instances first, then add new instances',
+                    value: ResizeStrategyType.DownScaleOldFirst
```
Comment (rameshx) [inline src/modules/75-cd/components/PipelineSteps/TasCanaryAppSetup/TasCanaryAppSetupWidget.tsx:184]: add labels to strings

**Review by rameshx — COMMENTED** [review]
Add all labels/placeholders to strings.

File: src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TasBasicAppSetupInputSet.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            multiTypeProps={{
+              expressions,
+              allowableTypes: [MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME]
```
Comment (Sagar-Rajput7) [inline src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TasBasicAppSetupInputSet.tsx:80]: please check if this is expected, as this is for runtime form.

File: src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TasBasicAppSetupInputSet.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            }}
+            label={getString('cd.steps.tas.additionalRoutes')}
+            enableConfigureOptions
```
Comment (Sagar-Rajput7) [inline src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TasBasicAppSetupInputSet.tsx:83]: same as above

File: src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TasBasicAppSetupInputSet.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      )}
+
+      {getMultiTypeFromValue((template?.spec as any)?.tempRoutes) === MultiTypeInputType.RUNTIME && (
+        <div className={cx(stepCss.formGroup, stepCss.md)}>
+          <FormMultiTypeKVTagInput
+            name={`${path}.spec.tempRoutes`}
+            tagsProps={{ placeholder: 'Type and press enter to add a route' }}
+            multiTypeProps={{
+              expressions,
+              allowableTypes: [MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME]
```
Comment (Sagar-Rajput7) [inline src/modules/75-cd/components/PipelineSteps/TASBasicAppSetupStep/TasBasicAppSetupInputSet.tsx:96]: Same as above

File: src/modules/75-cd/components/PipelineSteps/TasCanaryAppSetup/TasCanaryAppSetup.tsx [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+          Object.assign(errors, err)
+        }
+      }
+    }
+
+    /* istanbul ignore else */
+    if (
+      getMultiTypeFromValue(template?.spec?.existingVersionToKeep) === MultiTypeInputType.RUNTIME &&
+      isRequired &&
+      checkEmptyOrNegative(data?.spec?.existingVersionToKeep)
```
Comment (Sagar-Rajput7) [inline src/modules/75-cd/components/PipelineSteps/TasCanaryAppSetup/TasCanaryAppSetup.tsx:167]: **Nit:** Check if additional validation is required here, and also for BGAppSetup

**Review by Sagar-Rajput7 — DISMISSED** [review]
Nit: I see some of the fields and parts are common between these 3 steps and also some shared with Elastigroup steps also, so maybe we can sync together sometime in the future and think of making them generic or extract some to commonplace if possible.

**Comment by PratyushGar** [comment]
> Nit: I see some of the fields and parts are common between these 3 steps and also some shared with Elastigroup steps also, so maybe we can sync together sometime in the future and think of making them generic or extract some to commonplace if possible.

Sure, let's sync sometime and plan to make them generic.

File: src/modules/75-cd/components/PipelineSteps/PipelineStepsUtil.ts [hunk 9]
```diff
@@ -336,12 +336,12 @@ export const shouldFetchFieldData = (fieldList: string[]) => {
...
 }
 
-export const checkEmptyOrNegative = (value: any): boolean => /* istanbul ignore next */ {
+export const checkEmptyOrNegative = (value: any, minimumCount = 0): boolean => /* istanbul ignore next */ {
   if (typeof value === 'string') {
     return isEmpty(value)
   }
   if (typeof value === 'number') {
-    return value < 0
+    return value < minimumCount
```
Comment (rameshx) [inline src/modules/75-cd/components/PipelineSteps/PipelineStepsUtil.ts:344]: Should this be renamed as `checkEmptyOrLessThan`? 
The arg `minimumCount` can be just `minimum`/`minimumValue`.

**Review by Sagar-Rajput7 — DISMISSED** [review]

File: src/modules/75-cd/components/PipelineSteps/PipelineStepsUtil.ts (hunk 9 above)
Comment (PratyushGar) [reply src/modules/75-cd/components/PipelineSteps/PipelineStepsUtil.ts]: Agreed, thanks. Done

**Review by Sagar-Rajput7 — APPROVED** [review]

**Review by rameshx — APPROVED** [review]

**Pull request merged**